File: elements-of-programming-interviews/2023/python/rectangle_intersection.py
import collections

from test_framework import generic_test
from test_framework.test_failure import PropertyName

Rect = collections.namedtuple('Rect', ('x', 'y', 'width', 'height'))


def left(rect):
    return rect.x


def right(rect):
    return rect.x + rect.width


def down(rect):
    return rect.y


def up(rect):
    return rect.y + rect.height


def interval_overlap(x, l, r):
    return l <= x <= r


def line_overlap(a1, a2, b1, b2):
    return interval_overlap(a1, b1, b2) or \
        interval_overlap(a2, b1, b2) or \
        interval_overlap(b1, a1, a2) or \
        interval_overlap(b2, a1, a2)


def intersect_rectangle(r1: Rect, r2: Rect) -> Rect:
    if not (line_overlap(left(r1), right(r1), left(r2), right(r2)) and
            line_overlap(down(r1), up(r1), down(r2), up(r2))):
        return Rect(0, 0, -1, -1)

    x_vals = sorted([left(r1), right(r1), left(r2), right(r2)])
    y_vals = sorted([up(r1), down(r1), up(r2), down(r2)])
    return Rect(x_vals[1], y_vals[1], x_vals[2] - x_vals[1], y_vals[2] - y_vals[1])


def intersect_rectangle_wrapper(r1, r2):
    return intersect_rectangle(Rect(*r1), Rect(*r2))


def res_printer(prop, value):
    def fmt(x):
        return [x[0], x[1], x[2], x[3]] if x else None

    if prop in (PropertyName.EXPECTED, PropertyName.RESULT):
        return fmt(value)
    else:
        return value


if __name__ == '__main__':
    """
    r1 = Rect(76, 9, 12, 14)
    r2 = Rect(20, 1, 62, 60)
    # print(interval_overlap(up(r1), down(r2), up(r2)))
    """
    exit(
        generic_test.generic_test_main('rectangle_intersection.py',
                                       'rectangle_intersection.tsv',
                                       intersect_rectangle_wrapper,
                                       res_printer=res_printer))

File: elements-of-programming-interviews/2023/cpp/rectangle_intersection.cc
#include <tuple>

#include "test_framework/fmt_print.h"
#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"
struct Rect {
  // x and y are bottom left corner
  int x, y, width, height;
};

bool operator==(const Rect& r1, const Rect& r2) {
  return std::tie(r1.x, r1.y, r1.width, r1.height) ==
         std::tie(r2.x, r2.y, r2.width, r2.height);
}

bool pointIntervalIntersect(const int p, const int l, const int r) {
  // p in [l,r]?
  return l <= p && p <= r;
}

bool cornerOverlap1D(const int aLeft, const int aRight, const int bLeft,
                     const int bRight) {
  return pointIntervalIntersect(bLeft, aLeft, aRight) ||
         pointIntervalIntersect(bRight, aLeft, aRight) ||
         pointIntervalIntersect(aLeft, bLeft, bRight) ||
         pointIntervalIntersect(aRight, bLeft, bRight);
}

Rect IntersectRectangle(const Rect& r1, const Rect& r2) {
  int r1Left = r1.x;
  int r1Down = r1.y;
  int r1Right = r1.x + r1.width;
  int r1Up = r1.y + r1.height;

  int r2Left = r2.x;
  int r2Down = r2.y;
  int r2Right = r2.x + r2.width;
  int r2Up = r2.y + r2.height;

  if (!(cornerOverlap1D(r1Left, r1Right, r2Left, r2Right) &&
        cornerOverlap1D(r1Down, r1Up, r2Down, r2Up))) {
    return {0, 0, -1, -1};
  }
  std::vector<int> yVals{r1Up, r1Down, r2Up, r2Down};
  std::sort(yVals.begin(), yVals.end());
  std::vector<int> xVals{r1Left, r1Right, r2Left, r2Right};
  std::sort(xVals.begin(), xVals.end());

  return {xVals[1], yVals[1], xVals[2] - xVals[1], yVals[2] - yVals[1]};
}

namespace test_framework {
template <>
struct SerializationTrait<Rect> : UserSerTrait<Rect, int, int, int, int> {
  static std::vector<std::string> GetMetricNames(const std::string& arg_name) {
    return {FmtStr("height({})", arg_name), FmtStr("width({})", arg_name)};
  }

  static std::vector<int> GetMetrics(const Rect& x) {
    return {x.height, x.width};
  }
};
}  // namespace test_framework

std::ostream& operator<<(std::ostream& out, const Rect& r) {
  return PrintTo(out, std::make_tuple(r.x, r.y, r.width, r.height));
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"r1", "r2"};
  return GenericTestMain(args, "rectangle_intersection.cc",
                         "rectangle_intersection.tsv", &IntersectRectangle,
                         DefaultComparator{}, param_names);
}
